### PR TITLE
Retry: Fix permission denied error for web apps links from releases page

### PR DESCRIPTION
**Pull request by orangejenny**
https://manage.dimagi.com/default.asp?278241

Re-applies and fixes https://github.com/dimagi/commcare-hq/pull/21052 also taking https://github.com/dimagi/commcare-hq/pull/20865/ into account.

Also removes the deploy popup's web apps link for apps that don't have web apps enabled (in the current/canonical app), which will for real get a permission denied error.

@esoergel you're the most familiar with this saga
code buddies @millerdev / @emord

### CODE REVIEW

**Comment by esoergel** [comment]
Build failure looks real:
```Running "mocha:cloudcare" (mocha) task
Testing: http://localhost:8000/mocha/cloudcare/
Warning: SyntaxError: Parse error
 Use --force to continue.
Aborted due to warnings.
```

File: corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/api.js [hunk 1]
```diff
@@ -18,87 +18,91 @@ FormplayerFrontend.module("Menus", function (Menus, FormplayerFrontend, Backbone
...
-                        FormplayerFrontend.trigger(
-                            'showError',
-                            Formplayer.Errors.LOCK_TIMEOUT_ERROR
-                        );
-                    } else {
+            $.when(FormplayerFrontend.request("appselect:apps")).done(function (appCollection) {
+                if (!params.preview) {
+                    // Make sure the user has access to the app
+                    if (!appCollection.find(function(app) {
+                        return app.id === params.appId || app.get('copy_of') === params.copyOf;
```
Comment (esoergel) [inline corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/api.js]: Can you explain this one a little?  `appCollection` is a set of the most recent app builds, right?  Are there scenarios where `app.id === params.copyOf` or `app.copy_of === params.appId`?

I'm wondering if this should just check if there's anything in the set of all `app.id`s and `app.copy_of`s that matches either `params.appId` or `params.copyOf`.  That does sound a little heavy-handed, though.

**Review by esoergel — COMMENTED** [review]
I'd ignore whitespace when reviewing: https://github.com/dimagi/commcare-hq/pull/21194/files?w=1

**Review by millerdev — APPROVED** [review]
Don't feel like I have enough context here to give proper approval. Didn't see obvious problems in the code.

Aside: way easier to review with [w=1](https://github.com/dimagi/commcare-hq/pull/21194/files?w=1)

File: corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/api.js (hunk 1 above)
Comment (orangejenny) [reply corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/api.js]: Yep, `appCollection` is the most recent app builds, created by [get_web_apps_available_to_user](https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/cloudcare/views.py#L99) and used to render the set of app links that you see on the web apps home page. So `app.id` is always going to be a build id, and `app.copy_of` will always have the canonical app id.

`params.appId` is always going to be populated, and it'll always be a build id. For the links on the web apps home page, `params.appId` is always going to match one of the app ids in `appCollection`. For these links, `params.copyOf` doesn't need to be provided. For the links from the releases page, `params.appId` is still a build id, but it's likely not the latest build. These links now provide `params.copyOf` as a fallback.

So basically the first half of the conditional is for web apps home page links and the second half is for release page links.

File: corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/api.js [hunk 2]
```diff
@@ -18,87 +18,91 @@ FormplayerFrontend.module("Menus", function (Menus, FormplayerFrontend, Backbone
...
-                    }
-                },
-                error: function (_, response) {
-                    if (response.status === 423) {
-                        FormplayerFrontend.trigger(
-                            'showError',
-                            Formplayer.Errors.LOCK_TIMEOUT_ERROR
-                        );
-                    } else {
+            $.when(FormplayerFrontend.request("appselect:apps")).done(function (appCollection) {
```
[the remark this answers is not in the document]
Comment (esoergel) [reply corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/api.js]: Perfect, thanks for the investigation!

File: corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/api.js (hunk 1 above)
Comment (esoergel) [reply corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/api.js]: Thanks, that makes sense.  What about viewing an un-built, "current" application in webapps or live preview?  Will that have `params.appID` set to the canonical HQ app ID (ie `app.copy_of`) and `params.copyOf` set to null?

Comment (orangejenny) [reply corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/api.js]: No - you never really view an unbuilt app in web apps, even if set to view the current app (we make a build behind the scenes if one doesn't exist).

**Review by esoergel — APPROVED** [review]

**Review by esoergel — APPROVED** [review]

**Review by esoergel — APPROVED** [review]
